src/normalizer.ts

```diff
@@ -165,90 +330,104 @@ class V1SchemaNormalizer {
     }
   }
 
-  querySchemaRef(ref: string, location: string): Schema {
+  querySchemaRef(ref: string): Schema | null {
     const parts = ref.split('/')
-    if (parts[0] !== '#') throw new Error("Not a valid ref " + ref);
-    if (parts[1] !== 'components') throw new Error("Not a valid ref " + ref);
-    if (parts[2] !== 'schemas') throw new Error("Not a valid ref " + ref);
-    const name = parts[3];
+    if (parts[0] !== '#' || parts[1] !== 'components' || parts[2] !== 'schemas') {
+      this.recordError("Not a valid ref " + ref);
+      return null;
+    }
 
+    const name = parts[3];
     const s = this.schemas[name]
     if (!s) {
       const availableSchemas = Object.keys(this.schemas).join(', ')
-      throw new Error(`invalid reference ${ref}. Cannot find schema ${name}. Options are: ${availableSchemas}`);
+      this.recordError(`Invalid reference ${ref}. Cannot find schema ${name}. Options are: [${availableSchemas}]`);
+      return null;
     }
 
     return s
   }
 
-  // Recursively derive and annotate types
   annotateType(s: any): XtpNormalizedType | undefined {
     if (!s || typeof s !== 'object' || Array.isArray(s)) return undefined
-    if (s.xtpType) return s.xtpType // no need to recalculate
+    if (s.xtpType) return s.xtpType
 
-    // we can assume this is an object type
-    // if it has type = 'object' or has properties present
     if ((s.type && s.type === 'object') ||
       (s.properties && s.properties.length > 0)) {
 
-      const properties: XtpNormalizedType[] = []
-      for (const pname in s.properties!) {
-        const p = s.properties[pname]
-        const t = this.annotateType(p)!
-        p.xtpType = t
-        properties.push(t)
-      }
-
-      // TODO remove this legacy code
-      // we need to derive old type here
       s.type = 'object'
 
-      return new ObjectType(s.name!, properties, s)
+      const properties: XtpNormalizedType[] = []
+      if (s.properties) {
+        this.location.push('properties');
+        for (const pname in s.properties) {
+          const p = s.properties[pname]
+          this.location.push(p.name ?? pname);
+          const t = this.annotateType(p)
+          if (t) {
+            p.xtpType = t
+            properties.push(t)
+          }
+          this.location.pop();
+        }
```

Comment:
Having a little trouble understanding what this is doing. For example, when would `t` be falsey here?

Reply:
annotateType can return `undefined` in a few cases